controllers/nstemplatetier/nstemplatetier_controller.go

```diff
@@ -49,13 +53,187 @@ func (r *Reconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.
+	if tierTemplate.Spec.TemplateObjects == nil {
+		_, ok := nsTmplTier.Status.Revisions[tierTemplate.GetName()]
+		if !ok {
+			return true, tierTemplate.GetName(), nil
+		}
+		// nothing to update
+		return false, "", nil
+	}
+
+	if tierTemplateRevisionName, found := nsTmplTier.Status.Revisions[tierTemplate.GetName()]; found {
+		logger.Info("TTR set in the status.revisions for tiertemplate", "tierTemplate.Name", tierTemplate.GetName(), "ttr.Name", tierTemplateRevisionName)
+		var tierTemplateRevision toolchainv1alpha1.TierTemplateRevision
+		if err := r.Client.Get(ctx, types.NamespacedName{Namespace: nsTmplTier.GetNamespace(), Name: tierTemplateRevisionName}, &tierTemplateRevision); err != nil {
+			if errors.IsNotFound(err) {
+				// no tierTemplateRevision CR was found,
+				logger.Info("TTR CR not found", "tierTemplateRevision.Name", tierTemplateRevisionName)
+				// let's create one
+				ttrName, err := r.createNewTierTemplateRevision(ctx, nsTmplTier, tierTemplate)
+				if err != nil {
+					return false, "", err
+				}
+				// new ttr created
+				return true, ttrName, nil
+			} else {
+				// something wrong happened
+				return false, "", err
+			}
+		}
+		// TODO compare TierTemplate content with TTR content
+		// if the TierTemplate has changes we need to create new TTR
+	} else {
+		// no revision was set for this TierTemplate CR, let's create a TTR for it
+		ttrName, err := r.createNewTierTemplateRevision(ctx, nsTmplTier, tierTemplate)
+		if err != nil {
+			return false, ttrName, err
+		}
+		// new ttr created
+		return true, ttrName, nil
+	}
+	// nothing changed
+	return false, "", nil
+}
+
+func (r *Reconciler) createNewTierTemplateRevision(ctx context.Context, nsTmplTier *toolchainv1alpha1.NSTemplateTier, tierTemplate *toolchainv1alpha1.TierTemplate) (string, error) {
+	ttr := NewTTR(tierTemplate, nsTmplTier)
+	ttr, err := r.createTTR(ctx, ttr, tierTemplate)
+	if err != nil {
+		// something went wrong while creating new ttr
+		return "", err
+	}
+	return ttr.GetName(), nil
+}
+
+// getNSTemplateTierRefs returns a list with all the refs from the NSTemplateTier
+func getNSTemplateTierRefs(tmplTier *toolchainv1alpha1.NSTemplateTier) []string {
+	var refs []string
+	for _, ns := range tmplTier.Spec.Namespaces {
+		refs = append(refs, ns.TemplateRef)
+	}
+	if tmplTier.Spec.ClusterResources != nil {
+		refs = append(refs, tmplTier.Spec.ClusterResources.TemplateRef)
+	}
+
+	roles := make([]string, 0, len(tmplTier.Spec.SpaceRoles))
+	for r := range tmplTier.Spec.SpaceRoles {
+		roles = append(roles, r)
+	}
+	for _, r := range roles {
+		refs = append(refs, tmplTier.Spec.SpaceRoles[r].TemplateRef)
+	}
+	return refs
+}
+
+func (r *Reconciler) createTTR(ctx context.Context, ttr *toolchainv1alpha1.TierTemplateRevision, tmplTier *toolchainv1alpha1.TierTemplate) (*toolchainv1alpha1.TierTemplateRevision, error) {
+	err := r.Client.Create(ctx, ttr)
+	if err != nil && !errors.IsAlreadyExists(err) {
```

Comment:
we shouldn't really face any conflicts when the name of the TTR CR is generated

Reply:
we should not - it's a zombie from my _"copy paste driven development"_ 😆 🤦‍♂️ .

Reply:
But should we keep it? Just in case we would like to change the naming strategy and go away from the generated name, so that we don't have to remember to add this check back.

Reply:
Yeah, it's fine to keep it as it is, it won't cause any issues :+1:

Reply:
after reading: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/

> The server may generate a name when generateName is provided instead of name in a resource create request. When generateName is used, the provided value is used as a name prefix, which server appends a generated suffix to. Even though the name is generated, it may conflict with existing names resulting in a HTTP 409 response. This became far less likely to happen in Kubernetes v1.31 and later, since the server will make up to 8 attempt to generate a unique name before returning a HTTP 409 response.

I would rather drop it so we ensure that a new TTR CR is created
```suggestion
	if err != nil {
```

Reply:
make sense! Thanks!